### PR TITLE
feat (topsites): #2274 Pinning

### LINKED ISSUE #2274: Edit Top Sites: implement pinning
From #1987:  UI for the panel allows for pinning of Top Site cards

This should be a link/button on the top right of the tile that appears on hover.

screen shot 2017-03-20 at 11 52 10 am

^ Pinned tiles should have an indicator icon on the lower left corner.

We should reuse the pref used by the existing about:newtab or migrate users to the new system.

### PR DESCRIPTION
**Pull request by rlr**
Fixes #2274

### CODE REVIEW

File: addon/Feeds/TopSitesFeed.js [hunk 1]
```diff
@@ -10,26 +10,60 @@ const getScreenshot = require("addon/lib/getScreenshot");
...
   }
+  pinLink(action) {
+    this.pinnedLinks.pin(action.data.site, action.data.index);
+  }
+  unpinLink(action) {
+    this.pinnedLinks.unpin(action.data.site);
+  }
+  sortLinks(links, pinned) {
+    // Seperate out the pinned from the rest
+    let pinnedLinks = links.filter(link => this.pinnedLinks.isPinned(link));
```
Comment (ncloudioj) [inline addon/Feeds/TopSitesFeed.js]: nit: looks like the `links` was formed as "[pinnedLinks, nonPinnedFrecentLinks]". Can we just slice it into two sub-arrays instead of filtering twice here?

File: common/selectors/selectAndDedupe.js [hunk 2]
```diff
@@ -14,7 +16,21 @@ dedupe.defaults.createKey = site => site.cache_key || site.hostname || site.url;
...
-    let sites = options.defaults ? options.sites.concat(options.defaults) : options.sites;
+    let sites = options.sites;
+
+    if (options.defaults) {
+      // Dedupe the defaults first.
+      let defaults = dedupe.group([sites, options.defaults])[1];
+      // Fill in any empty slots with default sites.
+      sites.forEach((site, i) => {
+        if (!site) {
+          sites[i] = defaults.shift();
```
Comment (ncloudioj) [inline common/selectors/selectAndDedupe.js]: Any chance that `defaults.shift()` could return an `undefined`, i.e. running out of the `defaults`? Imaging a user has pinned two links at slot 1 and 12 in the "showMore" experiment, then he clears the whole history. In that case the default links (as currently 6 tiles) won't be enough to populate those 10 slots.

File: addon/Feeds/TopSitesFeed.js (hunk 1 above)
Comment (rlr) [inline addon/Feeds/TopSitesFeed.js]: good point. 👍

File: common/selectors/selectAndDedupe.js (hunk 2 above)
Comment (rlr) [inline common/selectors/selectAndDedupe.js]: yes, this is possible. Maybe we show the empty placeholder tiles for that case? Otherwise it will show the pin for slot 12 at slot 8 or something like that.

Comment (ncloudioj) [inline common/selectors/selectAndDedupe.js]: Sounds good. I think the placeholders will solve this issue.

File: addon/Feeds/TopSitesFeed.js [hunk 3]
```diff
@@ -10,26 +10,60 @@ const getScreenshot = require("addon/lib/getScreenshot");
...
+    return sortedLinks;
+  }
   getData() {
     return Task.spawn(function*() {
       const experiments = this.store.getState().Experiments.values;
 
-      let links;
-      // Get links from places
-      links = yield PlacesProvider.links.getTopFrecentSites();
+      // Get pinned links
```
Comment (ncloudioj) [inline addon/Feeds/TopSitesFeed.js]: What you think for this change? `getTopFrecentSites` already take a `limit` as one of the arguments. We can just add a new argument "excludedLinks". That also somewhat simplifies the code below.

Comment (ncloudioj) [inline addon/Feeds/TopSitesFeed.js]: Or do all those tweaks in a separate PR if needed. Your call :)

Comment (rlr) [inline addon/Feeds/TopSitesFeed.js]: Yeah, I think I'll hold off on adding the `excludedLinks` argument to the query since those queries are being commited to m-c. We should probably do the change over there once it lands.

File: content-src/components/NewTabPage/NewTabPage.js [hunk 4]
```diff
@@ -33,9 +33,9 @@ const NewTabPage = React.createClass({
         const topSites = this.props.TopSites.rows;
         stats.topsitesSize = topSites.length;
         topSites.forEach(row => {
-          if (row.screenshot) {
+          if (row && row.screenshot) {
```
Comment (ncloudioj) [inline content-src/components/NewTabPage/NewTabPage.js]: nice guard 👍

**Comment by as-pine-proxy** [comment]
[Treeherder: 16a133b816727452a06a5471acd543556787dbdf](https://treeherder.mozilla.org/#/jobs?repo=pine&revision=16a133b816727452a06a5471acd543556787dbdf)
